dill/logger.py

```diff
@@ -0,0 +1,276 @@
+        except AttributeError:
+            encoding = locale.getpreferredencoding()
+        try:
+            encoding = codecs.lookup(encoding).name
+        except LookupError:
+            self.is_utf8 = False
+        else:
+            self.is_utf8 = (encoding == codecs.lookup('utf-8').name)
+    def format(self, record):
+        fields = {'prefix': "", 'suffix': ""}
+        if getattr(record, 'depth', 0) > 0:
+            if record.msg.startswith("#"):
+                prefix = (record.depth - 1)*"│" + "└"
+            elif record.depth == 1:
+                prefix = "┬"
+            else:
+                prefix = (record.depth - 2)*"│" + "├┬"
+            if not self.is_utf8:
+                prefix = prefix.translate(ASCII_MAP) + "-"
+            fields['prefix'] = prefix + " "
+        if hasattr(record, 'size'):
+            # Show object size in human-redable form.
+            power = int(math.log(record.size, 2)) // 10
+            size = record.size >> power*10
+            fields['suffix'] = " [%d %sB]" % (size, "KMGTP"[power] + "i" if power else "")
+        vars(record).update(fields)
+        return super().format(record)
+
+logger = logging.getLogger('dill')
+adapter = TraceAdapter(logger)
+stderr_handler = logging.StreamHandler()
+adapter.addHandler(stderr_handler)
+
+def trace(arg: Union[bool, str, os.PathLike] = None, *, mode: str = 'a') -> NoReturn:
+    """print a trace through the stack when pickling; useful for debugging
+    
+    With a single boolean argument, enable or disable the tracing.
+    
+    Example usage:
+
+        >>> import dill
+        >>> dill.detect.trace(True)
+        >>> dill.dump_session()
+
+    Alternatively, ``trace()`` can be used as a context manager. With no
+    arguments, it just takes care of restoring the tracing state on exit.
+    A file name and a file mode may be specitfied to redirect the tracing
+    output in the ``with`` block context. A log function is yielded by the
+    manager so the user can write extra information to the file.
+
+    Example usage:
+
+        >>> from dill import detect
+        >>> D = {'a': 42, 'b': {'x': None}}
+        >>> with detect.trace():
+        >>>     dumps(D)
+        ┬ D2: <dict object at 0x7f2721804800>
+        ├┬ D2: <dict object at 0x7f27217f5c40>
+        │└ # D2 [8 B]
+        └ # D2 [22 B]
+        >>> squared = lambda x: x**2
+        >>> with detect.trace('output.txt', mode='w') as log:
+        >>>     log("> D = %r", D)
+        >>>     dumps(D)
+        >>>     log("> squared = %r", squared)
+        >>>     dumps(squared)
+
+    Arguments:
+        arg: a boolean value or an optional ``PathLike`` object for the context manager
+        mode: mode string for ``open()`` if a file name is passed as the first argument
+    """
+    if not isinstance(arg, bool):
+        return _trace(filename=arg, mode=mode)
+    logger.setLevel(logging.INFO if arg else logging.WARNING)
+
+@contextmanager
+def _trace(filename, mode):
```

Comment:
I think it is more robust if it takes a file handle, instead of a filename.  You could then do like pandas... and open the file if given the filename.

Reply:
Are you ok with the behavior that happens if someone gives `dill.detect.trace('test.log')` without using a context?

Reply:
_First comment:_ Taking an open stream is good if the user already have an open file, otherwise they would need to use two nested `with` statements or a `with` statement with two context managers (which I don't know if most users are aware of) like:
```python
with open('/tmp/trace', 'w') as file, dill.detect.trace(file) as log:
    # code
```

I think it's better to cover both use cases then.

---

_Second comment:_ No, not really... The first, class based, implementation avoided side effects if not called from a `with` statement. I'll revert to that.